server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -151,14 +152,26 @@ public synchronized void onNewCheckpoint(final ReplicationCheckpoint receivedChe
         } else {
             latestReceivedCheckpoint.put(replicaShard.shardId(), receivedCheckpoint);
         }
-        if (onGoingReplications.isShardReplicating(replicaShard.shardId())) {
-            logger.trace(
-                () -> new ParameterizedMessage(
-                    "Ignoring new replication checkpoint - shard is currently replicating to checkpoint {}",
-                    replicaShard.getLatestReplicationCheckpoint()
-                )
-            );
-            return;
+        Optional<SegmentReplicationTarget> ongoingReplicationTarget = onGoingReplications.getOngoingReplicationTarget(
+            replicaShard.shardId()
+        );
+        if (ongoingReplicationTarget.isPresent()) {
+            final SegmentReplicationTarget target = ongoingReplicationTarget.get();
+            if (target.getCheckpoint().getPrimaryTerm() < receivedCheckpoint.getPrimaryTerm()) {
+                logger.info(
+                    "Cancelling ongoing replication from old primary with primary term {}",
+                    target.getCheckpoint().getPrimaryTerm()
+                );
+                target.cancel("Cancelling stuck target after new primary");
```

Comment:
This should be invoking cancel through onGoingSegmentReplications cancel method so that it is removed from the collection.

Reply:
Updated